src/pybind/mgr/cephadm/ssh.py

```diff
@@ -0,0 +1,136 @@
+import asyncssh
+import asyncssh.logging
+import logging
+import os
+from io import StringIO
+from shlex import quote
+from typing import TYPE_CHECKING, Optional, List, Tuple
+from orchestrator import OrchestratorError
+
+if TYPE_CHECKING:
+    from cephadm.module import CephadmOrchestrator
+    from asyncssh import SSHClientConnection
+
+logger = logging.getLogger(__name__)
+
+
+class SSHConnection:
+
+    cons = {}
+
+    def __init__(self, mgr: "CephadmOrchestrator"):
+        self.mgr: "CephadmOrchestrator" = mgr
+
+    async def _remote_connection(self,
+                                 host: str,
+                                 addr: Optional[str] = None,
+                                 ) -> "SSHClientConnection":
+        if not SSHConnection.cons.get(host):
+            if not addr and host in self.mgr.inventory:
+                addr = self.mgr.inventory.get_addr(host)
+
+            self.mgr.offline_hosts_remove(host)
+
+            if not addr:
+                raise OrchestratorError("host address is empty")
+
+            assert self.mgr.ssh_user
+            n = self.mgr.ssh_user + '@' + addr
+            logger.debug("Opening connection to {} with ssh options '{}'".format(
+                n, self.mgr._ssh_options))
+
+            ssh_logger = logging.getLogger()
+            asyncssh.logging.set_log_level(logging.DEBUG)
+            log_string = StringIO()
+            ch = logging.StreamHandler(log_string)
+            ch.setLevel(logging.DEBUG)
+            ssh_logger.addHandler(ch)
+
+            try:
+                conn = await asyncssh.connect(addr, username=self.mgr.ssh_user, client_keys=[self.mgr.tkey.name], known_hosts=None, config=[self.mgr.ssh_config_fname])
+                SSHConnection.cons[host] = conn
+            except OSError as e:
+                log_content = log_string.getvalue()
+                log_string.flush()
+                logger.debug(log_content)
+                msg = f"Can't communicate with remote host `{addr}`, possibly because python3 is not installed there. {str(e)}" + '\n' + f'Log: {log_content}'
+                ssh_logger.removeHandler(ch)
+                raise OrchestratorError(msg)
+            except asyncssh.Error as e:
+                o = self._check_host(host)
+                if o is not None:
+                    log_content = log_string.getvalue()
+                    log_string.flush()
+                    logger.debug(log_content)
+                    msg = f'Failed to connect to {host} ({addr}) due to: {str(e)}' + '\n' + f'Log: {log_content}'
+                else:
+                    msg = f'Failed to connect to {host} ({addr}) due to: {str(e)}'
+                ssh_logger.removeHandler(ch)
```

Comment:
first thing to get this in a proper shape would be to create a method

```python
@contextmanager
def retirect_log()
    ssh_logger = logging.getLogger()
    asyncssh.logging.set_log_level(logging.DEBUG)
    log_string = StringIO()
    ch = logging.StreamHandler(log_string)
    ch.setLevel(logging.DEBUG)
    ssh_logger.addHandler(ch)
    
    try:
        yield
    except ???:
        log_content = log_string.getvalue()
        log_string.flush()
        logger.debug(log_content)
    finally:
        ssh_logger.removeHandler(ch)
```
I think that would help us getting a better understanding of when we can use this log-redirection